include/swift/AST/DiagnosticEngine.h

```diff
@@ -741,6 +746,35 @@ namespace swift {
       return diagnosticDocumentationPath;
     }
 
+    std::unique_ptr<diag::LocalizationProducer>
+    setDefaultLocalization(std::string defaultLocale, std::string path) {
+      defaultLocale = "en";
+      path = "/Volumes/Extreme/swift-source/build/Ninja-DebugAssert/swift-macosx-x86_64/share/swift/diagnostics";
+      assert(!path.empty());
+      assert(!defaultLocale.empty());
```

Comment:
I think this method shouldn't take any argument. I'll also need to find out how can I get the path for the current build like it's done in `CompilerInvocations.cpp`... I just did that for testing purposes on my machine.

Reply:
It seems like what you want here is for `DiagnosticEngine` constructor to accept a path to localization directory and locale ("default" could be hardcoded to `en` always).

Reply:
Also `defaultLocalization` doesn't have to be dynamically allocated since it would always be there and that would avoid pointer chasing when accessing it.

Reply:
I've been thinking about how I'd pass the default localization path to the `DiagnosticEngine`'s constructor for quite some time now. 
So, how can I get a path from [`CompilerInvocations::setMainExecutablePath`](https://github.com/apple/swift/blob/master/lib/Frontend/CompilerInvocation.cpp#L62-L82) to use in the DiagnosticEngine's constructor?

As far as I know, I can't use [`CompilerInstance::setUpDiagnosticOptions`](https://github.com/apple/swift/blob/master/lib/Frontend/Frontend.cpp#L398) to pass the default localization path to the `DiagnosticEngine` (like we do with `LocalizationPath`).

I was thinking *maybe* we might use some sort of a global variable that will always have the `MainExecutablePath` for diagnostics, but I'm not sure. 

WDYT I should do?

Reply:
`DiagnosticEngine` is attached to a `CompilerInstance` and passed around e.g. to the `CompilerInvocation`, `ASTContext` etc. Take a look at `swift::performFrontend` method - it calls `getMainExecutable` - I think we could move it up and pass result as an argument to a constructor of `CompilerInstance` and use it to initialize `DiagnosticEngine` field attached to it (https://github.com/apple/swift/blob/master/lib/FrontendTool/FrontendTool.cpp#L2500). This also eliminates the need to do anything in `setMainExecutablePath`.